apiextensions/v1alpha1/types_compatibilityrequirement.go

```diff
@@ -0,0 +1,368 @@
+}
+
+// CompatibilityRequirementSpec is the specification of the desired behavior of the Compatibility Requirement.
+type CompatibilityRequirementSpec struct {
+	// compatibilitySchema defines the schema used by
+	// customResourceDefinitionSchemaValidation and objectSchemaValidation.
+	// This field is required.
+	// +required
+	CompatibilitySchema CompatibilitySchema `json:"compatibilitySchema,omitempty,omitzero"`
+
+	// customResourceDefinitionSchemaValidation ensures that updates to the
+	// installed CRD are compatible with this compatibility requirement. If not
+	// specified, admission of the target CRD will not be validated.
+	// This field is optional.
+	// +optional
+	CustomResourceDefinitionSchemaValidation CustomResourceDefinitionSchemaValidation `json:"customResourceDefinitionSchemaValidation,omitempty,omitzero"`
+
+	// objectSchemaValidation ensures that matching resources conform to
+	// compatibilitySchema. If not specified, admission of matching resources
+	// will not be validated.
+	// This field is optional.
+	// +optional
+	ObjectSchemaValidation ObjectSchemaValidation `json:"objectSchemaValidation,omitempty,omitzero"`
+}
+
+// CRDDataType indicates the type of the CRD data.
+// +kubebuilder:validation:Enum=YAML
+type CRDDataType string
+
+const (
+	// CRDDataTypeYAML indicates that the CRD data is in YAML format.
+	CRDDataTypeYAML CRDDataType = "YAML"
+)
+
+// CRDData contains the complete definition of a CRD.
+type CRDData struct {
+	// type indicates the type of the CRD data. The only supported type is "YAML".
+	// This field is required.
+	// +required
+	Type CRDDataType `json:"type,omitempty"`
+
+	// data contains the complete definition of the CRD. This field must be in
+	// the format specified by the type field. It may not be longer than 1572864
+	// characters.
+	// This field is required.
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=1572864
+	// +required
+	Data string `json:"data,omitempty"`
+}
+
+// APIVersionSelectionType specifies a method for automatically selecting a
+// set of API versions to require.
+// +kubebuilder:validation:Enum=StorageOnly;AllServed
+type APIVersionSelectionType string
+
+const (
+	APIVersionSetTypeStorageOnly APIVersionSelectionType = "StorageOnly"
+	APIVersionSetTypeAllServed   APIVersionSelectionType = "AllServed"
+)
+
+// APIVersionString is a string representing a kubernetes API version.
+// +kubebuilder:validation:MinLength=1
+// +kubebuilder:validation:MaxLength=63
+// +kubebuilder:validation:XValidation:rule="!format.dns1035Label().validate(self).hasValue()",message="version must be a valid DNS 1035 label"
+type APIVersionString string
+
+// APIVersions specifies a set of API versions of a CRD.
+// +kubebuilder:validation:XValidation:rule="self.defaultSelection != 'AllServed' || !has(self.additionalVersions)",message="additionalVersions may not be defined when defaultSelection is 'AllServed'"
```

Comment:
What if I want to include a version that is in the schema definition but is not served?